File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.cpp
#include <string>
#include <vector>
#include <math.h>
#include <string.h>
#include <queue>
#include <iostream>

using namespace std;

bool adj[101][101];
bool visited[101];


int bfs(int start, int n) {
    visited[start] = true;
    queue<int> q;
    q.push(start);
    int result = 0;
    while(!q.empty()) {
        int size = q.size();
        while(size--) {
            result++;
            int node = q.front();
            q.pop();
            for (int next = 1; next <= n; next++) {
                if (adj[node][next] && !visited[next]) {
                    visited[next] = true;
                    q.push(next);
                }
            }
        }
    }
    return result;
}

int solution(int n, vector<vector<int>> wires) {
    int answer = 1e9;
    for (int i = 0; i < wires.size(); i++) {
        int u = wires[i][0];
        int v = wires[i][1];
        adj[u][v] = true;
        adj[v][u] = true;
    }
    
    for (int i = 0; i < wires.size(); i++) {
        int u = wires[i][0];
        int v = wires[i][1];
        adj[u][v] = false;
        adj[v][u] = false;
        int a = bfs(u, n);
        memset(visited, 0, sizeof(visited));
        
        int b = bfs(v, n);
        memset(visited, 0, sizeof(visited));   
        answer = min(answer, abs(a-b));
        adj[u][v] = true;
        adj[v][u] = true;
    }

    return answer;
}

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    
    int a, b, c;
    
    public int solution(int n, int[][] wires) {
        int answer = Integer.MAX_VALUE;
        
        Map<Integer, List<Integer>> tree = new HashMap<>();
        for (int i = 1; i <= n; i++) {
            tree.put(i, new ArrayList<>());
        }
        
        for (int[] w : wires) {
            Integer u = w[0];
            Integer v = w[1];
            tree.get(u).add(v);
            tree.get(v).add(u);
        }
        
        for (int[] w : wires) {
            Integer u = w[0];
            Integer v = w[1];
            tree.get(u).remove(v);
            tree.get(v).remove(u);
            
            boolean[] visited = new boolean[n + 1];
            a = 0; b = 0;
            for (int i = 1; i <= n; i++) {
                if (!visited[i]) {
                    dfs(tree, i, visited);
                    break;
                }
            }
            c = 1;
            for (int i = 1; i <= n; i++) {
                if (!visited[i]) {
                    dfs(tree, i, visited);
                    break;
                }
            }
            c = 0;

            answer = Math.min(answer, Math.abs(a - b));
            
            tree.get(u).add(v);
            tree.get(v).add(u);
        }
        return answer;
    }
    
    void dfs(Map<Integer, List<Integer>> graph, int here, boolean[] visited) {
        visited[here] = true;
        
        if (c == 0) a++;
        else b++;
        for (int i = 0; i < graph.get(here).size(); i++) {
            int next = graph.get(here).get(i);
            if (!visited[next]) {
                dfs(graph, next, visited);
            }
        }
        
    }
}